jvector-base/src/main/java/io/github/jbellis/jvector/graph/disk/ParallelGraphWriter.java

```diff
@@ -0,0 +1,362 @@
+     * memory issues when processing very large graphs.
+     *
+     * @param ordinalMapper maps between old and new ordinals
+     * @param inlineFeatures the inline features to write
+     * @param featureStateSuppliers suppliers for feature state
+     * @param baseOffset the file offset where L0 records start
+     * @throws IOException if an IO error occurs
+     */
+    public void writeL0Records(OrdinalMapper ordinalMapper,
+                               List<Feature> inlineFeatures,
+                               Map<FeatureId, IntFunction<Feature.State>> featureStateSuppliers,
+                               long baseOffset) throws IOException {
+        // Calculate optimal buffer size based on available heap memory
+        int bufferSize = calculateOptimalBufferSize();
+        int maxOrdinal = ordinalMapper.maxOrdinal();
+        List<Future<NodeRecordTask.Result>> futures = new ArrayList<>(bufferSize);
+
+        // Build and write records in chunks to manage memory usage
+        for (int newOrdinal = 0; newOrdinal <= maxOrdinal; newOrdinal++) {
+            final int ordinal = newOrdinal;
+            final long fileOffset = baseOffset + (long) ordinal * recordSize;
+
+            // Submit task to build this record
+            Future<NodeRecordTask.Result> future = executor.submit(() -> {
+                var view = viewPerThread.get();
+                var buffer = bufferPerThread.get();
+
+                var task = new NodeRecordTask(
+                        ordinal,
+                        ordinalMapper,
+                        graph,
+                        view,
+                        inlineFeatures,
+                        featureStateSuppliers,
+                        recordSize,
+                        fileOffset,
+                        buffer
+                );
+
+                return task.call();
+            });
+
+            futures.add(future);
+
+            // When buffer is full, write this batch and clear for next batch
+            if (futures.size() >= bufferSize) {
+                writeRecordsAsync(futures);
+                futures.clear();
+            }
+        }
+
+        // Write any remaining records
+        if (!futures.isEmpty()) {
+            writeRecordsAsync(futures);
+        }
+    }
+
+    /**
+     * Calculates the optimal buffer size based on available heap memory.
+     * Uses a conservative approach to prevent memory exhaustion while maintaining
+     * good parallelism and throughput.
+     * <p>
+     * The calculation considers:
+     * - Current heap usage and maximum heap size
+     * - Estimated memory per record (data + Future/Result overhead)
+     * - Conservative allocation (20% of available memory)
+     * - Reasonable bounds (min 100, max 1000000 records)
+     *
+     * @return the number of records to buffer before writing
+     */
+    private int calculateOptimalBufferSize() {
```

Comment:
Should we take graph properties to this calculation to consider alignment with device blocks?